data/trie/branchNode.go

```diff
@@ -388,61 +446,68 @@ func (bn *branchNode) getNext(key []byte, db data.DBWriteCacher) (node, []byte,
 	return bn.children[childPos], key, nil
 }
 
-func (bn *branchNode) insert(n *leafNode, db data.DBWriteCacher) (bool, node, [][]byte, error) {
+func (bn *branchNode) insert(n *leafNode, db data.DBWriteCacher) (node, [][]byte, error) {
 	emptyHashes := make([][]byte, 0)
 	err := bn.isEmptyOrNil()
 	if err != nil {
-		return false, nil, emptyHashes, fmt.Errorf("insert error %w", err)
+		return nil, emptyHashes, fmt.Errorf("insert error %w", err)
 	}
-	if len(n.Key) == 0 {
-		return false, nil, emptyHashes, ErrValueTooShort
+
+	insertedKey := n.Key
+	if len(insertedKey) == 0 {
+		return nil, emptyHashes, ErrValueTooShort
 	}
-	childPos := n.Key[firstByte]
+	childPos := insertedKey[firstByte]
 	if childPosOutOfRange(childPos) {
-		return false, nil, emptyHashes, ErrChildPosOutOfRange
+		return nil, emptyHashes, ErrChildPosOutOfRange
 	}
-	n.Key = n.Key[1:]
+
+	n.Key = insertedKey[1:]
 	err = resolveIfCollapsed(bn, childPos, db)
 	if err != nil {
-		return false, nil, emptyHashes, err
+		return nil, emptyHashes, err
 	}
 
-	if bn.children[childPos] != nil {
-		var dirty bool
-		var newNode node
-		var oldHashes [][]byte
-
-		dirty, newNode, oldHashes, err = bn.children[childPos].insert(n, db)
-		if !dirty || err != nil {
-			return false, bn, emptyHashes, err
-		}
-
-		if !bn.dirty {
-			oldHashes = append(oldHashes, bn.hash)
-		}
-
-		bn.children[childPos] = newNode
-		bn.dirty = dirty
-		if dirty {
-			bn.hash = nil
-		}
-		return true, bn, oldHashes, nil
+	if bn.children[childPos] == nil {
+		return bn.insertOnNilChild(n, childPos)
 	}
 
+	return bn.insertOnExistingChild(n, childPos, db)
+}
+
+func (bn *branchNode) insertOnNilChild(n *leafNode, childPos byte) (node, [][]byte, error) {
 	newLn, err := newLeafNode(n.Key, n.Value, bn.marsh, bn.hasher)
 	if err != nil {
-		return false, nil, emptyHashes, err
+		return nil, [][]byte{}, err
 	}
-	bn.children[childPos] = newLn
 
-	oldHash := make([][]byte, 0)
+	modifiedHashes := make([][]byte, 0)
+	modifiedHashes = bn.modifyNodeAfterInsert(modifiedHashes, childPos, newLn)
+
+	return bn, modifiedHashes, nil
+}
+
+func (bn *branchNode) insertOnExistingChild(n *leafNode, childPos byte, db data.DBWriteCacher) (node, [][]byte, error) {
+	newNode, modifiedHashes, err := bn.children[childPos].insert(n, db)
+	if check.IfNil(newNode) || err != nil {
```

Comment:
Could you have nil newNode and nil err?
then you end up with nil err

if it cannot happen then you don't need the check.IfNil right?

Reply:
There is a case where you can have a nil newNode and a nil err: if you try to insert the same value at the same key (leafNode.go L233). It is not an error, but there is no newNode to return. The insert() return is treated in particiaMerkleTrie.go on L132 - L138.